Disable register hooks on startup via property

samrocketman opened the pull request:
This change provides an admin the ability to better control how GhprbTrigger initializes.

- Convert existing system properties to use standards from Jenkins core.  This is necessary to sanitize user input to ensure that system properties are properly handled.  This also integrates with Jenkins system logging when a user accidentally misconfigures a property.
- Add a system property to disable registering GitHub webhooks on startup.  This assumes that hooks were already properly registered when an admin or user created the Jenkins job.
- Implements recommendations from #535

This change provides the following system properties:

- `org.jenkinsci.plugins.ghprb.GhprbTrigger.poolSize=N` where `N` is an integer defining number of threads to use for registering webhooks on startup.
- `org.jenkinsci.plugins.ghprb.GhprbTrigger.disableRegisterOnStartup=N` where `N` is boolean `true` or `false`.

Additional notes:

- Jenkins 2 is now a requirement.
- Updated all dependencies to resolve building against Jenkins 2.
- Added pinned dependencies to resolve build issues with the maven enforcer plugin.
- Added more debug logging.

samrocketman commented:
Build currently fails.  I'm working on updating the dependency chains for utilizing the SystemProperties class.

samrocketman commented:
Build has been fixed.  Noteworthy updates:

- Jenkins 2 is now the minimum Jenkins core requirement.
- Updates were made to support Jenkins 2 as the minimum required version.
- See commit message for additional details.

This change is ready for review and possible merge.

bjoernhaeuser commented on src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java:
```diff
@@ -70,7 +72,7 @@
     private static final Logger logger = Logger.getLogger(GhprbTrigger.class.getName());
 
     private static final ExecutorService pool = Executors.newFixedThreadPool(
```
Do you have tests, which are baking this change? We are running Jenkins with 32 Cores, which would mean 32 parallel requests to github, all waiting for github's api to return. This just feels like hammering with a very big hammer on a very tiny nail :)

samrocketman replied:
I'm not sure how to write tests for this use case.

The thread pool is configurable because I understand one size doesn't fit all.  For my setup, we have GitHub Enterprise with thousands of jobs so I would probably make the count 64 (or just disable it on startup entirely).

bjoernhaeuser replied:
I meant did you test what this change will do for you?

It totally makes sense to make this configurable, just arguing that defaulting to the amount of cores could lead to some very large threadpools :)

samrocketman replied:
Yeah, but I don't run a 32 core master server and tend to use large machines as agents.  If a large thread pool is a concern then I don't mind keeping the same values as before and just updating how the property is actually read.

samrocketman commented:
@bjoernhaeuser please continue to scrutinize.  I squashed my changes and preserved original behavior.  I mostly am concerned about proper checking of user input via system properties and the ability to disable hook registration on Jenkins startup.

I have added more debug logging so I can continue to test this.  It's not ready to merge until I finish up my testing.

chr0n1x commented:
👍  to the disable flag! 😄

bjoernhaeuser commented on src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java:
```diff
@@ -275,7 +286,16 @@ public void start(Job<?, ?> project, boolean newInstance) {
+            //disableRegisterOnStartup system property allows an admin to
+            //disable this behavior.  This assumes hooks are already properly
+            //registered when a job is first created.
+            boolean disable_register_on_startup = SystemProperties.getBoolean(getClass().getName() + ".disableRegisterOnStartup", false);
```
I suggest to move this to class level (next to pool definition) and to rename it to `disabletoRegisterOnStartUp`, that seems to be more in line with the rest

samrocketman replied:
Amended commit with your feedback.

samrocketman commented:
Hold off on merging.  This currently requires the latest version of Jenkins.  I'm going to work on supporting lower versions of plugins, first.

samrocketman commented:
`jenkins.util.SystemProperties` was introduced in Jenkins 2.4.  That should be the new minimum version of Jenkins allowed and not the latest non-LTS version.  I'll update this another day this week.

samrocketman commented:
# Screenshot of logs during testing

![screenshot from 2017-08-05 03-48-55](https://user-images.githubusercontent.com/875669/28994991-8d5740fa-7991-11e7-9b39-d2492c2ef7af.png)

# Ready for review and possible merge

I've finally finished up this PR.  I feel it is ready for production use.  Please review @benpatterson @bjoernhaeuser @chr0n1x 

Tested on Jenkins 2.60.2.

samrocketman commented:
@benpatterson since this has had a pretty thorough review; and I don't have merge permission; do you mind merging this?

benpatterson commented:
Hmm I'll see if I can look at repo permissions... But indeed, looks good and merging.
